In [8]:
import pandas as pd

# Load dataset
df = pd.read_csv("station_data_dataverse.csv")

# --- Step 1: Fix timestamp strings with known year error ---
def fix_year_safe(ts):
    if pd.isna(ts): return ts
    return ts.replace("0014", "2014")

df['created'] = pd.to_datetime(df['created'].apply(fix_year_safe), errors='coerce')
df['ended'] = pd.to_datetime(df['ended'].apply(fix_year_safe), errors='coerce')

# --- Step 2: Convert chargeTimeHrs to numeric and recover missing durations ---
df['chargeTimeHrs'] = pd.to_numeric(df['chargeTimeHrs'], errors='coerce')
df.loc[df['chargeTimeHrs'].isna() & df['created'].notna() & df['ended'].notna(), 'chargeTimeHrs'] = \
    (df['ended'] - df['created']).dt.total_seconds() / 3600

# --- Step 3: Clean and fill 'distance' ---
df['distance'] = pd.to_numeric(df['distance'], errors='coerce')
df['distance'].fillna(df['distance'].median(), inplace=True)

# --- Step 4: Flag outliers and extract derived features ---
df['isOutlier'] = df['chargeTimeHrs'] > 12
df['hour'] = df['created'].dt.hour
df['isWeekend'] = df['created'].dt.weekday >= 5
df['dayOfWeek'] = df['created'].dt.day_name()

# --- Step 5: Flag rows with both timestamps missing (retain them) ---
df['timestamp_missing'] = df['created'].isna() & df['ended'].isna()

# --- Step 6: Save cleaned dataset ---
df.to_csv("Cleaned_EV_Charging_Sessions.csv", index=False)

print("✅ Data cleaning complete. Saved as 'Cleaned_EV_Charging_Sessions.csv'.")
print(f"Rows retained: {df.shape[0]}")
print(f"Rows with missing timestamps flagged: {df['timestamp_missing'].sum()}")


✅ Data cleaning complete. Saved as 'Cleaned_EV_Charging_Sessions.csv'.
Rows retained: 3395
Rows with missing timestamps flagged: 3372


EV Charger Pattern Analysis – Data Cleaning Progress

1. Dataset Loaded

File: station_data_dataverse.csv

Total rows: 3,395

2. Timestamp Handling

Replaced malformed '0014' with '2014' in timestamp fields

Converted created and ended to proper datetime format

Recalculated chargeTimeHrs where missing using ended - created

3. Flagged Missing Timestamps

Created a new column: timestamp_missing

Flagged rows where both created and ended are missing (NaT)

Flagged: 3,372 rows

4. Distance Cleaning

Converted distance to numeric format

Imputed missing values using the median distance

5. Outlier Detection

Marked sessions with chargeTimeHrs > 12 hours as outliers

6. Feature Extraction

Extracted hour from created timestamp

Derived isWeekend and dayOfWeek for user behavior analysis

Time-based features are available for 23 valid timestamp rows

7. Output Summary

Cleaned dataset saved as: Cleaned_EV_Charging_Sessions.csv

Total rows retained: 3,395

Rows with valid timestamps: 23

Dataset is fully usable for duration-based clustering; time-based clustering requires filtering


In [10]:
!pip install pymongo

   ---------------------------------------- 0.0/846.0 kB ? eta -:--:--
   ---------------------------------------- 10.2/846.0 kB ? eta -:--:--
   - ------------------------------------- 41.0/846.0 kB 495.5 kB/s eta 0:00:02
   ------- -------------------------------- 163.8/846.0 kB 1.2 MB/s eta 0:00:01
   -------------------------- ------------- 563.2/846.0 kB 3.2 MB/s eta 0:00:01
   ---------------------------------------  839.7/846.0 kB 4.4 MB/s eta 0:00:01
   ---------------------------------------  839.7/846.0 kB 4.4 MB/s eta 0:00:01
   ---------------------------------------- 846.0/846.0 kB 3.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/313.6 kB ? eta -:--:--
   --------------------------------------  307.2/313.6 kB 18.6 MB/s eta 0:00:01
   ---------------------------------------- 313.6/313.6 kB 9.8 MB/s eta 0:00:00


In [23]:
from pymongo import MongoClient
import pandas as pd

# --- Step 1: Connect to MongoDB ---
client = MongoClient("mongodb+srv://EVAT:EVAT123@cluster0.5axoq.mongodb.net/EVAT?retryWrites=true&w=majority&appName=Cluster0")
db = client['EVAT']  # Replace with your actual database name if different

# --- Step 2: Read your dataset ---
df = pd.read_csv('Cleaned_EV_Charging_Sessions.csv')  # Replace with your actual CSV file name

# --- Step 3: Choose or create collection ---
collection = db['Cleaned_EV_Charging_Sessions']  # Replace with your target collection name

# --- Step 4: Convert DataFrame to dictionary and upload ---
data = df.to_dict(orient='records')
collection.insert_many(data)

print(f"✅ Uploaded {len(data)} records to the collection.")

# --- 📝 Step 5: Verification ---
sample = collection.find_one()
print("🔍 Sample document inserted:")
print(sample)


✅ Uploaded 3395 records to the collection.
🔍 Sample document inserted:
{'_id': ObjectId('6802f8255d0e47f0f72d52fa')}
